llvm/lib/Transforms/IPO/SampleProfile.cpp

```diff
@@ -478,20 +505,31 @@ class SampleProfileMatcher {
                      std::map<LineLocation, StringRef> &IRAnchors);
   void findProfileAnchors(
       const FunctionSamples &FS,
-      std::map<LineLocation, std::unordered_set<FunctionId>>
-          &ProfileAnchors);
-  void countMismatchedSamples(const FunctionSamples &FS);
-  void countProfileMismatches(
-      const Function &F, const FunctionSamples &FS,
-      const std::map<LineLocation, StringRef> &IRAnchors,
-      const std::map<LineLocation, std::unordered_set<FunctionId>>
-          &ProfileAnchors);
-  void countProfileCallsiteMismatches(
-      const FunctionSamples &FS,
-      const std::map<LineLocation, StringRef> &IRAnchors,
+      std::map<LineLocation, std::unordered_set<FunctionId>> &ProfileAnchors);
+  // Record the callsite match states for profile staleness report, the result
+  // is saved in FuncCallsiteMatchStates.
+  void recordCallsiteMatchStates(
+      const Function &F, const std::map<LineLocation, StringRef> &IRAnchors,
       const std::map<LineLocation, std::unordered_set<FunctionId>>
           &ProfileAnchors,
-      uint64_t &FuncMismatchedCallsites, uint64_t &FuncProfiledCallsites);
+      const LocToLocMap *IRToProfileLocationMap);
+
+  bool isMismatchState(const enum MatchState &State) {
```

Comment:
This is only called during counting (post-match), so we should not expect any Initial* state. 

Can we assert somewhere during counting that we do not have any initial* state? If we have that means `recordCallsiteMatchStates` may have a bug.

Reply:
I found that we still can get the Initial* state; because currently the fuzzy matching is only on when there is checksum mismatch, but the counting and reporting is for all non-import function, so for those checksum matched function, their state are on the  Initial* state.

Reply:
If checksum actually matches, do we expect any InitialMismatch from such functions? 

Or in other words, if checksum matches, there is no state change to track, do we actually still need to run `recordCallsiteMatchStates`? Can we just count how many samples, callsites and callsite samples in those functions and assume they all match?

Reply:
> If checksum actually matches, do we expect any InitialMismatch from such functions?
> 
> Or in other words, if checksum matches, there is no state change to track, do we actually still need to run `recordCallsiteMatchStates`? Can we just count how many samples, callsites and callsite samples in those functions and assume they all match?

It could happen, I saw in our dashboard, even the checksum mismatched samples is zero, there are a few numbers of callsite mismatches(e.g. Stale function samples: 0.00%(773327/65978370982), Stale functions: 0.01%(6/42787) , Stale callsite samples: 0.29%(194149265/65978370982), Stale callsite samples after matching: 0.29%(194318114/65978370982), Stale callsites: 5.95%(20780/349307) ,  Stale callsites after matching:  5.95%(20781/349307))

I checked that it's mostly from the ctor alias optimization, it's one kind of "function-rename"(in this case it's not from user change), on the other hand, the function-renaming from user change is also in this category.  Checking on checksum matches can help detect this issue. 

Though this ctor alias functions are all very cold functions, it's maybe fine to ignore this if this "false positives" (this is more like a future support)  is noisy. 

Any thoughts?

Reply:
Just to make sure I understand this, so this is because we compute hash on CFG (ignoring callee names). If a callee is renamed, even if checksum matches for the caller, we can have call site mismatch due to callee renaming (without caller CFG change). Is that the issue? 

I was hoping that we don't need to pay the cost of recording and computing stats if there is no mismatch. But if that's not possible, current change is fine. 

It begs another questions though... should checksum actually capture callee renaming? like if callee renamed, should caller's checksum change?

Reply:
>Just to make sure I understand this, so this is because we compute hash on CFG (ignoring callee names). If a callee is renamed, even if checksum matches for the caller, we can have call site mismatch due to callee renaming (without caller CFG change). Is that the issue?

Exactly. 

> I was hoping that we don't need to pay the cost of recording and computing stats if there is no mismatch. But if that's not possible, current change is fine.
> It begs another questions though... should checksum actually capture callee renaming? like if callee renamed, should caller's checksum change?

This is a good question. We do use some callsite info for computing checksum: the total number of callsites. If we add/remove one more callsite, it's still captured as checksum changes. so it's not strict enough. 

If we really want to capture the name changes, we can encode the name of callsite into the checksum.  Then again, similar to the "unstable IR" issue, there could be mismatches from internal compiler but not from user code. This ctor alias is one case. We will run stale matching for this kind of changes. That said, it's doable, just we needs more changes to recognize : "this is not from user renaming, let's only run renaming opt but not fuzzy matching opt". the current fuzzy matching may not work well on this ctor alias cases. We need more check.

Reply:
Yes I was thinking about capturing callee names in checksum. But we don't have to decide now. Let's deal with that when we work on handling renames.

Reply:
There should be no other unresolved comments, I'm going to push this change. I really appreciate all the discussions and suggestions!